joblib/disk.py

```diff
@@ -111,14 +112,20 @@ def delete_folder(folder_path, onerror=None):
             # if the error is raised again, fail
             err_count = 0
             while True:
+                files = os.listdir(folder_path)
                 try:
-                    shutil.rmtree(folder_path, False, None)
-                    break
+                    if len(files) == 0 or allow_non_empty:
+                        shutil.rmtree(folder_path, False, None)
+                        util.debug(
+                            "Sucessfully deleted {}".format(folder_path))
+                        break
+                    else:
+                        raise OSError
```

Comment:
```suggestion
                        msg = f"Expected empty folder {folder_path} but got {len(files)} files."
                        raise OSError()
```

to make the error message more explicit in case this is the last exception to be raised when `err_count` has reached its limit.